x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java

```diff
@@ -63,24 +64,26 @@ public Weight doCache(Weight weight, QueryCachingPolicy policy) {
      */
     static boolean cachingIsSafe(Weight weight, IndicesAccessControl.IndexAccessControl permissions) {
         // support caching for common queries, by inspecting the field
-        // TODO: If in the future there is a Query#extractFields() then we can do a better job
-        Set<String> fields = new HashSet<>();
         try {
-            FieldExtractor.extractFields(weight.getQuery(), fields);
-        } catch (UnsupportedOperationException ok) {
-            // we don't know how to safely extract the fields of this query, don't cache.
-            return false;
-        }
+            weight.getQuery().visit(new QueryVisitor() {
+                @Override
+                public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
+                    return this; // we want to use the same visitor for must_not clauses too
+                }
 
-        // we successfully extracted the set of fields: check each one
-        for (String field : fields) {
-            // don't cache any internal fields (e.g. _field_names), these are complicated.
-            if (field.startsWith("_") || permissions.getFieldPermissions().grantsAccessTo(field) == false) {
-                return false;
-            }
+                @Override
+                public boolean acceptField(String field) {
+                    // don't cache any internal fields (e.g. _field_names), these are complicated.
+                    if (field.startsWith("_") || permissions.getFieldPermissions().grantsAccessTo(field) == false) {
+                        throw new UnsupportedOperationException();
```

Comment:
nit: I've never used it myself, but I think this is a great case where we can throw an exception type for which the stack trace is not populated, such that we avoid that overhead, in cases like these where exception throwing works as a break out of a loop.

Reply:
agreed, I was even considering whether I should go out of my way to not throw exception here, throwing did not feel super natural.

Reply:
I've pushed https://github.com/elastic/elasticsearch/pull/93390/commits/[073586b](https://github.com/elastic/elasticsearch/pull/93390/commits/073586bfa2d8d8116d012b658175870ba48f6651)fa2d8d8116d012b658175870ba48f6651 to use a dedicated exception type that doesn't populate the stacktrace.